### PR TITLE
Experimental: Unix Perm add support for SyncManager's download

### PR DESCRIPTION
**Pull request by N-o-Z**
https://github.com/treeverse/dev/issues/356

### CODE REVIEW

File: pkg/local/unix_permissions.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+		}
+	}
+	return UnixPermissions{
+		unixOwnership: *defaultOwnership,
+		Mode:          os.FileMode(mode),
+	}
+}
+
+// getUnixPermissionFromStats - Get unix mode and ownership from object metadata, fallback to default permissions in case metadata doesn't exist
+func getUnixPermissionFromStats(stats apigen.ObjectStats) (*UnixPermissions, error) {
```
Comment (itaigilo) [inline pkg/local/unix_permissions.go:62]: `GetUnixPermissionFromStats`?

**Review by itaigilo — COMMENTED** [review]
Looks ok, from what I can understand.

A few comments, most are nitpicking, I guess.

File: pkg/local/unix_permissions.go [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+		defaultOwnership = &unixOwnership{
+			UID: os.Getuid(),
+			GID: os.Getgid(),
+		}
+	}
+	return UnixPermissions{
+		unixOwnership: *defaultOwnership,
+		Mode:          os.FileMode(mode),
+	}
+}
```
Comment (guy-har) [inline pkg/local/unix_permissions.go:59]: IIUC, this is called from many goroutines

File: pkg/local/unix_permissions.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	"strings"
+	"syscall"
+
+	"github.com/treeverse/lakefs/pkg/api/apigen"
+	"github.com/treeverse/lakefs/pkg/api/apiutil"
+	"github.com/treeverse/lakefs/pkg/uri"
+)
+
+const (
+	DefaultFileMask            = 0o100666
```
Comment (guy-har) [inline pkg/local/unix_permissions.go:15]: Can you document this

File: pkg/local/unix_permissions.go [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+		if ok {
+			// Unmarshal struct
+			if err := json.Unmarshal([]byte(unixPermissions), &permissions); err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return &permissions, nil
+}
```
Comment (guy-har) [inline pkg/local/unix_permissions.go:75]: Just a suggestion becuase I like it flat
```suggestion
func getUnixPermissionFromStats(stats apigen.ObjectStats) (*UnixPermissions, error) {
	permissions := getDefaultPermissions(strings.HasSuffix(stats.Path, uri.PathSeparator))
	if stats.Metadata == nil {
		return &permissions, nil
	}

	unixPermissions, ok := stats.Metadata.Get(UnixPermissionsMetadataKey)
	if !ok {
		return &permissions, nil
	}

	// Unmarshal struct
	if err := json.Unmarshal([]byte(unixPermissions), &permissions); err != nil {
		return nil, err
	}
	return &permissions, nil
}
}
```

**Review by guy-har — COMMENTED** [review]
Cool 😎

File: pkg/local/progress.go [hunk 5]
```diff
@@ -119,7 +119,7 @@ func NewProgressPool() *ProgressPool {
 	pw := progress.NewWriter()
 	pw.SetAutoStop(false) // important
 	pw.SetTrackerLength(progressTrackerLength)
-	pw.SetMessageWidth(progressTrackerWidth)
+	pw.SetMessageLength(progressTrackerWidth)
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/local/progress.go:122]: Upgrading the package, this method has become deprecated and needs to be replaced by the latter

File: pkg/local/sync.go [hunk 6]
```diff
@@ -158,86 +162,104 @@ func (s *SyncManager) download(ctx context.Context, rootPath string, remote *uri
 		_ = json.Unmarshal(statResp.Body, &httpErr)
 		return fmt.Errorf("(stat: HTTP %d, message: %s): %w", statResp.StatusCode(), httpErr.Message, ErrDownloadingFile)
 	}
+	ObjStat := *statResp.JSON200
 	// get mtime
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/local/sync.go:166]: Done

File: pkg/local/unix_permissions.go (hunk 1 above)
Comment (N-o-Z) [reply pkg/local/unix_permissions.go:62]: No need to make it public (at least for now)

File: pkg/local/sync.go [hunk 7]
```diff
@@ -158,86 +162,104 @@ func (s *SyncManager) download(ctx context.Context, rootPath string, remote *uri
...
+	mtimeSecs, err := getMtimeFromStats(ObjStat)
 	if err != nil {
 		return err
 	}
+	lastModified := time.Unix(mtimeSecs, 0)
 
-	if strings.HasSuffix(path, uri.PathSeparator) {
+	var unixPerm *UnixPermissions
+	if s.includePerm {
+		if unixPerm, err = getUnixPermissionFromStats(ObjStat); err != nil {
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply pkg/local/sync.go:175]: It is used to fail fast. We don't want to download the file and then realize we can't get the permissions

File: pkg/local/unix_permissions.go (hunk 4 above)
Comment (N-o-Z) [reply pkg/local/unix_permissions.go:75]: I find it harder to follow

File: pkg/local/sync.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (guy-har) [reply pkg/local/sync.go]: IIUC the only error that can return here is on jsln. unmarshal, am I missing something, or are we failing fast on that? 
Anyway please document that it's in order to fail fast

File: pkg/local/unix_permissions.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+	DefaultFilePermissions     = 0o100666
+	UnixPermissionsMetadataKey = apiutil.LakeFSMetadataPrefix + "unix-permissions"
+)
+
+var (
+	// umask - internal, init only once. Use only via getDefaultPermissions call
+	umask = -1
+	// defaultOwnership - internal, init only once. Use only via getDefaultPermissions call
+	defaultOwnership  *unixOwnership
+	getOwnershipMutex sync.Mutex
```
Comment (guy-har) [inline pkg/local/unix_permissions.go:26]: Another option would be to this functionality as methods of the sync manager and initiate it on new.

**Review by guy-har — APPROVED** [review]

File: pkg/local/unix_permissions.go (hunk 4 above)
Comment (guy-har) [reply pkg/local/unix_permissions.go:80]: 🤷‍♂️

**Pull request merged**